### DISCUSSION TITLE
How to customize user registration

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [OrchardCMS/OrchardCore#11240] [Q&A]
With [some discussion already](https://github.com/OrchardCMS/OrchardCore/discussions/9264), but what I am more interested in is the _out of the box_ `dev`/`author` experience. There appears to be a `Notify Me` button on the `Coming Soon` recipe, for instance, and an action asscoaited with that form. However, I do not see anywhere to inject assets, resources, credentials, etc, connecting the dots with an actual email send/receive process. How do we go about connecting that, let along troubleshooting? I was able to drudge up some old youtubes on the topic, but they are very dated, I think; certainly obsolete as compared/contrasted with contemporary `OrchardCore/CMS` versions. Thank you...

**Comment by mwpowellhtx** [comment]
So far I am finding the `Configuration/Features` _admin_ section... Have not investigated beyond that point, so insights are welcome, of course, but intuition tells me I check and enable the facets that I am interested in.

**Reply by Skrypt** [reply]
The emails are sent through a Workflow generally. So, each time that this form is posted it will be caught by the Workflow and then it will send a customized email to an email of your choice. In this case, it is not a user registration, it's just a FORM post with some data persisted in your database through a Workflow.

**Reply by mwpowellhtx** [reply]
> In this case, it is not a user registration, it's just a FORM post with some data persisted in your database through a Workflow.

Okay. How is it persisted, _'where'_, in what domain model? I can dive deeper, of course, but if there is some documentation, that'd be fantastic. Something like that, eventually, I think, should/could potentially connect the dots with the _identity/user/role credential_, I would expect. But probably that is a bit beyond its current scope.

Also, emails sent? How? Where? I gather there need to be _Email_, perhaps _SMTP_, _credentials_ established supporting that workflow? That is not immediately obvious from the workflow. Again, some notes about that would be fantastic.

i.e. configured resources, assets, supporting the workflow, as part of the `Configuration/Features`? Not 100% sure about that at the moment.

**Reply by Skrypt** [reply]
Here you have the documentation about the Workflow module : 
https://docs.orchardcore.net/en/dev/docs/reference/modules/Workflows/
Yes, of course, to send emails you need to enable the email feature in Orchard Core. Once enabled you will get a page in the Orchard Core configurations where you can set the SMTL credentials or simply use a folder for testing.

The Workflows can be parsed to create a Content Item of a specific Content Type. That way, you get a "normalized" data record. Else, each Workflow execution stores its own JSON metadata which includes the form post data. 

There is one method that I used to parse the form post data in javascript. This is why we use script expressions in the different workflow activities / events.

https://docs.orchardcore.net/en/dev/docs/reference/modules/Workflows/#javascript-functions-in-http-activities

Look at `deserializeRequestData()` which will return the request data as JSON for you to create a Content Item of a specific content type later on. Also, from the Email task you will be able to use something like : `{{ Workflow.Properties["ContactRequest"].Email | raw }}`

So basically your Workflow will need a starting "event" which is an "HTTP Request Event" set to capture "POST" events.
Then you will link this "HTTP Request Event" to a "Define Property" task which will use the `deserializeRequestData()` and set a property named "ContactRequest". Then an Email task linked to that "Define Property" task which will use Liquid in it's template to render the email : 

```liquid
<div>
Someone sent you a contact request from : 
{{ Workflow.Properties["ContactRequest"].Email | raw }}
</div>
```

**Comment by Skrypt** [comment]
Here is what the Workflow looks like in the "Coming Soon" recipe.

![image](https://user-images.githubusercontent.com/3228637/155019338-4e4f5bc1-7097-4af1-aad4-c85eaac8b558.png)

As you can see here we are using the "Bind form to model state" which will create server-side values to use directly per form field like here  "Email". It's doing the same magic thing as the MVC model binding.

Here, normally, the only thing required is to set the Email credentials : 

![image](https://user-images.githubusercontent.com/3228637/155019459-811426cf-bd43-4bb2-9697-0034db0c8975.png)

**Reply by mwpowellhtx** [reply]
Yep tracking with that much. Just curious what dots I need to connect; apparently _'those'_, for starters, depending what other insights might be there.

**Reply by Skrypt** [reply]
You could easily remove the "Create User" task there if you don't need to register them as users in your website database. I think, these "Coming Soon" forms should generally just send an email to the user and be kept inside an email list that can be retrieved for sending mass emails like _aka_ Mailchimp. Also, as you can see you can use scripting / Liquid inside Workflow tasks to get the form data values.

You can see another example here : 

https://docs.orchardcore.net/en/dev/docs/reference/modules/Forms/

**Reply by mwpowellhtx** [reply]
Well, true, there's _'that'_, but it is also interesting to connect them with the user management and a proper profile. Good to know the events however, thank you.

**Reply by mwpowellhtx** [reply]
So if I can recap what I think is going on.

1. User enters their contact details and clicks `Notify Me`
1. The form action is connected with the `User Registration` workflow
1. Which depends upon there being an Email resource configured (or folder based, whatever, ...)
1. And the workflow is processed accordingly

Ostensibly with some sort of event that we may subscribe to in order to connect contact details with user profiles.
